robot-server/robot_server/protocols/protocol_store.py

```diff
@@ -0,0 +1,119 @@
+"""Methods for saving and retrieving protocol files."""
+from dataclasses import dataclass
+from datetime import datetime
+from fastapi import UploadFile
+from logging import getLogger
+from pathlib import Path
+from typing import Dict, List, Sequence
+
+from opentrons.file_runner import ProtocolFileType
+
+log = getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class ProtocolStoreEntry:
+    """An entry in the session store, used to construct response models."""
+
+    protocol_id: str
+    protocol_type: ProtocolFileType
+    created_at: datetime
+    files: List[Path]
+
+
+class ProtocolNotFoundError(ValueError):
+    """Error raised when a protocol ID was not found in the store."""
+
+    def __init__(self, protocol_id: str) -> None:
+        """Initialize the error message from the missing ID."""
+        super().__init__(f"Protocol {protocol_id} was not found.")
+
+
+class ProtocolStore:
+    """Methods for storing and retrieving protocol files."""
+
+    def __init__(self, directory: Path) -> None:
+        """Initialize the ProtocolStore.
+
+        Arguments:
+            directory: Directory in which to place created files.
+        """
+        self._directory = directory
+        self._protocols_by_id: Dict[str, ProtocolStoreEntry] = {}
+
+    async def create(
+        self,
+        protocol_id: str,
+        created_at: datetime,
+        files: Sequence[UploadFile],
+    ) -> ProtocolStoreEntry:
+        """Add a protocol to the store."""
+        protocol_dir = self._get_protocol_dir(protocol_id)
+        protocol_dir.mkdir(parents=True, exist_ok=True)
```

Comment:
I understand `parents=True` to automatically create the root directory (`self._directory`), but why `exist_ok=True`? I'd expect it to be an error if I tried to do something that would accidentally partially overwrite an existing protocol.

(If it turns out that we want `create()` to be idempotent, we should document that. But I don't think that's something we want, at least at the store level.)

```suggestion
        # TODO(mc, 2021-06-02): check for protocol collision
        protocol_dir.mkdir(parents=True)
```

Reply:
`mkdir(exist_ok=True)` will no-op if the directory exists. That feels safer to me, and makes this logic easier to re-use when we get around to adding files to existing protocols